### DISCUSSION TITLE
print error

### DISCUSSION THREAD

**Discussion by starthlo** [gptme/gptme#139] [General]
![Screenshot_1](https://github.com/user-attachments/assets/dc8b4dc0-01d6-49c8-93e4-e36ecd0c5dab)

It appears that `rich.print` is not functioning as expected. Instead of displaying just the result of the command, it is also showing progress messages. 

![Screenshot_2](https://github.com/user-attachments/assets/5f2cff40-98ec-4a8f-a0f6-d3e645820830)

This command will ask us confirmation to uninstall `tmux`. As you know, we need to answer `yes` or `no`.
However, the confirmation prompt is not being displayed during the uninstallation process. We need to update the command and execute it again using sudo apt-get remove tmux.

**Comment by ErikBjare** [comment]
Yes, the stdout capturing will duplicate status messages. 

I think the correct way to handle that is to set appropriate env variables to make command-line apps not do that (like setting `DEBIAN_FRONTEND=noninteractive` in the case of apt), at least when gptme is run in non-interactive mode. But we would need something similar for `sudo` commands.

Although when gptme is run in interactive mode, we could perhaps allow the user to input, but seems difficult and not entirely sure its a good idea. Perhaps the agent should be allowed to input, but its hard to know when input is requested.

**Comment by starthlo** [comment]
I understand. How can we solve `rich.print` error - first screenshot?

**Comment by TimeToBuildBob** [comment]
Re-read this on **July 11, 2026** against the current `gptme` tree.

This thread is mixing **two different issues**:

1. The **first screenshot is not a `rich.print` crash**. In normal text mode, `gptme` can show its own progress/status output alongside captured command output. If you need machine-clean stdout today, use:
   ```bash
   gptme --non-interactive --output-format json ...
   ```
   That mode landed on **May 13, 2026** (`50422d121`), and stdout purity in JSON mode was tightened on **May 26, 2026** (`ed66a68c1`).

2. The **missing `sudo` prompt** in the second screenshot is the TTY side, not Rich. Current `gptme` has explicit interactive `sudo` handling in the shell tool; that was fixed on **February 20, 2026** (`e79ff6e9f`). For unattended package commands, the right pattern is still to make the command itself non-interactive, e.g. `DEBIAN_FRONTEND=noninteractive` plus apt flags like `-y`, and `sudo -n` if you do not want a password prompt.

I also re-ran the current regression slice today:

```bash
uv run pytest -q \
  tests/test_tools_shell.py::test_needs_tty_sudo_detection \
  tests/test_output_format.py \
  tests/test_acp_stdout.py
```

Result: **21 passed**.

So this looks stale rather than an open `rich.print` bug. If you can still reproduce it on a current release, please open a fresh issue with:
- `gptme --version`
- OS + terminal
- exact command
- whether it was interactive or `--non-interactive`

That will make the remaining surface reproducible instead of guessing from old screenshots.
